hw15-messagingsystems/src/main/java/ru/otus/hw15messagingsystems/messagesystem/MessageSystemImpl.java

```
@@ -0,0 +1,159 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Service("MessageSystem")
+public final class MessageSystemImpl implements MessageSystem {
+  private static final Logger logger = LoggerFactory.getLogger(MessageSystemImpl.class);
+  private static final int MESSAGE_QUEUE_SIZE = 100_000;
+  private static final int MSG_HANDLER_THREAD_LIMIT = 2;
+
+  private final AtomicBoolean runFlag = new AtomicBoolean(true);
+
+  private final Map<String, MsClient> clientMap = new ConcurrentHashMap<>();
+  private final BlockingQueue<Message> messageQueue = new ArrayBlockingQueue<>(MESSAGE_QUEUE_SIZE);
+
+  private Runnable disposeCallback;
+
+  private final ExecutorService msgProcessor = Executors.newSingleThreadExecutor(runnable -> {
+    Thread thread = new Thread(runnable);
+    thread.setName("msg-processor-thread");
+    return thread;
+  });
+
+  private final ExecutorService msgHandler = Executors.newFixedThreadPool(MSG_HANDLER_THREAD_LIMIT, new ThreadFactory() {
+    private final AtomicInteger threadNameSeq = new AtomicInteger(0);
+
+    @Override
+    public Thread newThread(Runnable runnable) {
+      Thread thread = new Thread(runnable);
+      thread.setName("msg-handler-thread-" + threadNameSeq.incrementAndGet());
+      return thread;
+    }
+  });
+
+  public MessageSystemImpl() {
+    start();
+  }
+
+  public MessageSystemImpl(boolean startProcessing) {
+    if (startProcessing) {
+      start();
+    }
+  }
+
+  @Override
+  public void start() {
+    msgProcessor.submit(this::msgProcessor);
+  }
+
+  @Override
+  public int currentQueueSize() {
+    return messageQueue.size();
+  }
+
+  @Override
+  public void addClient(MsClient msClient) {
+    logger.info("new client:{}", msClient.getName());
+    if (clientMap.containsKey(msClient.getName())) {
+      throw new IllegalArgumentException("Error. client: " + msClient.getName() + " already exists");
+    }
+    clientMap.put(msClient.getName(), msClient);
```

Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **65 and 68**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 If the key already exists, the code should throw an error at line: **66**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.